framework/aster-frame/src/vm/page_table/mod.rs

```diff
@@ -0,0 +1,400 @@
+    /// usize overflows, the caller should provide the index range of the root level pages
+    /// instead of the virtual address range.
+    pub(crate) fn make_shared_tables(&self, root_index: Range<usize>) {
+        let start = root_index.start;
+        debug_assert!(start >= nr_ptes_per_node::<C>() / 2);
+        debug_assert!(start < nr_ptes_per_node::<C>());
+        let end = root_index.end;
+        debug_assert!(end <= nr_ptes_per_node::<C>());
+        let mut root_frame = self.root_frame.lock();
+        for i in start..end {
+            let no_such_child = root_frame.child(i).is_none();
+            if no_such_child {
+                let frame = Arc::new(SpinLock::new(PageTableFrame::<E, C>::new()));
+                root_frame.map(
+                    i,
+                    Child::PageTable(frame),
+                    Some(PageProperty {
+                        flags: PageFlags::RWX,
+                        cache: CachePolicy::Writeback,
+                        priv_flags: PrivilegedPageFlags::GLOBAL,
+                    }),
+                    false,
+                )
+            }
+        }
+    }
+}
+
+impl<'a, M: PageTableMode, E: PageTableEntryTrait, C: PagingConstsTrait> PageTable<M, E, C>
+where
+    [(); nr_ptes_per_node::<C>()]:,
+    [(); C::NR_LEVELS]:,
+{
+    /// Create a new empty page table. Useful for the kernel page table and IOMMU page tables only.
+    pub(crate) fn empty() -> Self {
+        PageTable {
+            root_frame: Arc::new(SpinLock::new(PageTableFrame::<E, C>::new())),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// The physical address of the root page table.
+    pub(crate) fn root_paddr(&self) -> Paddr {
+        self.root_frame.lock().start_paddr()
+    }
+
+    pub(crate) unsafe fn map(
+        &self,
+        vaddr: &Range<Vaddr>,
+        paddr: &Range<Paddr>,
+        prop: PageProperty,
+    ) -> Result<(), PageTableError> {
+        self.cursor_mut(vaddr)?.map(MapOption::MapUntyped {
+            pa: paddr.start,
+            len: vaddr.len(),
+            prop,
+        });
+        Ok(())
+    }
+
+    pub(crate) unsafe fn unmap(&self, vaddr: &Range<Vaddr>) -> Result<(), PageTableError> {
+        self.cursor_mut(vaddr)?
+            .map(MapOption::Unmap { len: vaddr.len() });
+        Ok(())
+    }
+
+    pub(crate) fn protect(
+        &self,
+        vaddr: &Range<Vaddr>,
+        op: impl FnMut(&mut PageProperty),
+    ) -> Result<(), PageTableError> {
+        self.cursor_mut(vaddr)?
+            .protect(vaddr.len(), op, true)
+            .unwrap();
+        Ok(())
+    }
+
+    /// Query about the mapping of a single byte at the given virtual address.
+    pub(crate) fn query(&self, vaddr: Vaddr) -> Option<(Paddr, PageProperty)> {
+        // Safety: The root frame is a valid page table frame so the address is valid.
+        unsafe { page_walk::<E, C>(self.root_paddr(), vaddr) }
```

Comment:
I am considering performance issues since it is faster not to check locks. The MMU can't afford concurrent modifications while walking either. It is crucial for the framework user to ensure not modifying the page table while other threads can read from the address, so this should be satisfied.

However, we seems to have already messed up with it, even if this is implemented with the cursor. Maybe we could fix it in the future.

Reply:
Fix: I've already checked that a PTE is always written after the physical address (of the child node/frame) points to initialized objects. So as long as MMU walk holds this holds too.